src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCachePolicyProvider.cs

```diff
@@ -144,41 +162,55 @@ public virtual bool IsCachedEntryFresh(ResponseCacheContext context)
             var cachedControlHeaders = context.CachedResponseHeaders.CacheControl ?? EmptyCacheControl;
 
             // Add min-fresh requirements
-            if (context.RequestCacheControlHeaderValue.MinFresh.HasValue)
+            var minFresh = context.RequestCacheControlHeaderValue.MinFresh;
+            if (minFresh.HasValue)
             {
-                age += context.RequestCacheControlHeaderValue.MinFresh.Value;
+                age += minFresh.Value;
+                context.Logger.LogFreshnessMinFreshAdded(minFresh.Value);
             }
 
             // Validate shared max age, this overrides any max age settings for shared caches
-            if (age >= cachedControlHeaders.SharedMaxAge)
+            var sharedMaxAge = cachedControlHeaders.SharedMaxAge;
+            if (age >= sharedMaxAge)
             {
                 // shared max age implies must revalidate
+                context.Logger.LogFreshnessSharedMaxAgeExceeded(age, sharedMaxAge.Value);
                 return false;
             }
-            else if (!cachedControlHeaders.SharedMaxAge.HasValue)
+            else if (!sharedMaxAge.HasValue)
             {
+                var cachedMaxAge = cachedControlHeaders.MaxAge;
+                var requestMaxAge = context.RequestCacheControlHeaderValue.MaxAge;
+                var lowestMaxAge = cachedMaxAge < requestMaxAge ? cachedMaxAge : requestMaxAge ?? cachedMaxAge;
                 // Validate max age
-                if (age >= cachedControlHeaders.MaxAge || age >= context.RequestCacheControlHeaderValue.MaxAge)
+                if (age >= lowestMaxAge)
                 {
                     // Must revalidate
                     if (cachedControlHeaders.MustRevalidate)
                     {
+                        context.Logger.LogFreshnessMustRevalidate(age, lowestMaxAge.Value);
                         return false;
                     }
 
                     // Request allows stale values
-                    if (age < context.RequestCacheControlHeaderValue.MaxStaleLimit)
+                    var maxStaleLimit = context.RequestCacheControlHeaderValue.MaxStaleLimit;
+                    if (maxStaleLimit.HasValue && age - lowestMaxAge < maxStaleLimit)
```

Comment:
Misread the RFC previously, the stale limit is the difference between current age and the specified max-age.